Add secp256k1-sys with prefixed sources

stevenroose opened the pull request:
I'm adding a secp256k1-sys crate with just the FFI wrapper of the C libsecp.

And then I added a script (thanks @jonasnick) to prefix the secp sources with a rust-secp specific tag and a version tag, so that we don't suffer from naming conflicts anymore.

elichai commented:
idk about this.
but first of all you brought back all the `malloc` functions we removed in #130 
And even about the prefix, there are people who do want to compile this code as libsecp (i.e. bitcoin-consensus and stuff @TheBlueMatt is working on). although in theory LTO should take care of the duplicates.

Even if other people do want this i'll split it to ~3 PRs. because right now it's not auditable.
(i.e. one PR creating the *-sys crate, another move *only* another adds the prefix)

stevenroose commented:
>  but first of all you brought back all the malloc functions we removed in #130

That was an accident. Fixed that with patch files.

> Even if other people do want this i'll split it to ~3 PRs. because right now it's not auditable.

What's wrong with one PR with 3 commits? The creating the -sys crate and moving the dep is not possible to do in 2 commits without having a giant diff of first adding the entire depend folder a second time in another place and then removing the original in a second commit instead of having git treat them as a  move.

I can do that if that is favorable, but that was the reason I did the -sys crate together with migrating secp to -sys in the same commit.

elichai commented:
@jonasnick I forgot about reviewing commit by commit, I'll try that. Sorry.

Isn't the whole point of LTO is to remove code duplication? So if you link against two versions of libsecp but with different symbols then most of the duplication should disappear it you use LTO.

elichai commented on secp256k1-sys/depend/secp256k1/src/asm/field_10x26_arm.s:
```diff
@@ -27,16 +27,16 @@ Note:
 	.align	2
-	.global secp256k1_fe_mul_inner
-	.type	secp256k1_fe_mul_inner, %function
+	.global rustsecp256k1_v0_1_0_fe_mul_inner
```
is this function linked as a public function?

stevenroose replied:
No idea, it's changed by autoreplace.

jonasnick commented:
@elichai to be clear, are you aware that this PR contains a script `secp256k1-sys/vendor-libsecp.sh ` that just replaces all references of secp256k1 with `rustsecp256k1_v0_1_0`?

elichai commented:
@jonasnick well we still need to review that the script did it's job currently and then the person who ran it didn't change anything.

Is it possible in C to have private functions? or all functions are "public" so we need to mangle all their names? even inner functions

jonasnick commented:
> then the person who ran it didn't change anything.

You can verify the script then run it yourself. But okay.

Just renaming may introduce bugs that are not caught by the compiler. But that seems highly unlikely and will also be difficult to notice in manual review.

> even inner functions

Are you suggesting to have a more complicated script that only looks at the exported symbols? Why?

stevenroose commented on secp256k1-sys/depend/secp256k1/Makefile.am:
```diff
@@ -2,13 +2,13 @@ ACLOCAL_AMFLAGS = -I build-aux/m4
 
```
That is just the auto replace at work.

stevenroose commented:
@elichai so you can run the following command to verify the depends folder:
```
cd ./secp256k1-sys/
./vendor-libsecp.sh depend 0_1_0 143dc6e9ee31852a60321b23eea407d2006171da
```

The commit has is committed in the `./secp256k1-sys/depend/secp256k1-HEAD-revision.txt` file.

real-or-random commented on secp256k1-sys/vendor-libsecp.sh:
```diff
@@ -0,0 +1,67 @@
+patch "$DIR/include/secp256k1.h" "./secp256k1.h.patch"
+patch "$DIR/src/secp256k1.c" "./secp256k1.c.patch"
+patch "$DIR/src/scratch_impl.h" "./scratch_impl.h.patch"
+patch "$DIR/src/util.h" "./util.h.patch"
```
Would it be simpler to rename the Rust functions?

real-or-random replied:
Maybe add a comment why we remove the other functions.

stevenroose replied:
Apparently the existence of the functions breaks a WASM build. I have not verified that, but that's why they were removed in the first place.

elichai replied:
Yep. It your build target has no `malloc` symbol than it will fail linking.  (which some of the wasm targets has no malloc symbols

elichai commented:
@stevenroose I tried to link this against upstream and `ecdsa_signature_parse_der_lax` is missing.
I guess we should build that. because it's not an official part of upstream.

TheBlueMatt commented on Cargo.toml:
```diff
@@ -11,32 +11,32 @@ documentation = "https://docs.rs/secp256k1/"
+
+# Use this feature to not compile the bundled libsecp256k1 C symbols,
+# but use external ones.
+external-symbols = ["secp256k1-sys/external-symbols"]
```
Can you just *not* depend on secp256k1-sys at all if you build with external-symbols? pulling in a dummy crate as a dependency sucks.

stevenroose replied:
The crate is still defining the C interface and the FFI objects used in it. In fact it's not doing less than when built without `external-symbols`, it's just the build.rs construction is no longer utilized.

apoelstra replied:
At least we should add a scarier warning.

Also, as you noted in another channel, this has the potential to cause UB a user is using an old version of rust-secp, as well as this one with `external-symbols` enabled...so we can't depend on cargo failing if this is enabled on a cargo build.

stevenroose commented on secp256k1-sys/src/lib.rs:
```diff
@@ -300,6 +328,11 @@ pub unsafe extern "C" fn secp256k1_context_create(flags: c_uint) -> *mut Context
     secp256k1_context_preallocated_create(ptr as *mut c_void, flags)
 }
 
+#[cfg(all(feature = "std", not(feature = "dont_replace_c_symbols")))]
```
What do you mean with inline? I changed this to the external-symbols feature, so I'm not sure if your comment still applies.

stevenroose commented:
> If we anyway clone secp256k1, could we remove it entirely? Probably not because then you need to be online for building and we trust Github more? It's just a wild thought, please ignore me if this is stupid.

We're only cloning it when we revendor. That only needs to happen when updated to a new version. The sources are still required for being able to build.

stevenroose commented on secp256k1-sys/vendor-libsecp.sh:
```diff
@@ -0,0 +1,67 @@
+fi
+HEAD=$(git rev-parse HEAD)
+cd ..
+echo "\# This file was automatically created by $0" > ./secp256k1-HEAD-revision.txt
```
done

real-or-random commented on secp256k1-sys/README.md:
```diff
@@ -0,0 +1,34 @@
+
+The default build process is to build using the vendored libsecp256k1 sources in
+the depend folder. These sources are prefixed with a special
+rust-secp256k1-sys-specific prefix `rustsecp256k1_v1_2_3_`.
```
Maybe point out which version number this is supposed to be. Of the `rust-secp256k1` crate or of the `-sys` crate? (I assume both... :))

stevenroose replied:
No, it's just the -sys crate. The rust-secp crate version can freely move around and make breaking changes in the API without a need to update -sys. -sys should only update if the upstream sources update.

real-or-random commented on secp256k1-sys/vendor-libsecp.sh:
```diff
@@ -0,0 +1,68 @@
+# We need to make some source changes to the files.
+
+# To support compiling for WASM, we need to remove all methods that use malloc.
+# To compansate, the secp_context_create and _destroy methods are redefined in Rust.
```
```suggestion
# To compensate, the secp_context_create and _destroy methods are redefined in Rust.
```

elichai commented:
Semi related.
If anyone has any knowledge about elf/exe/(whatever mac has) symbols, please DM me (IRC etc.)
I want to build a tool that will be like an extension to the `cc` crate that will auto prefix/mangle the symbols and then we can easily integrate this with `bindgen` to generate valid easy  definitions even if the mangling is somewhat random. (already talked to the maintainers of these tools. `cc` prefer it will be external to it. `bindgen` says that he can easily support the link features)

Right now the only tool I found that can do this is `objcopy`. but it's part of binutils and obviously only work on elf.

apoelstra commented:
When we merge this let's be sure to get #162 in as well.

stevenroose commented:
Any updates on this?

elichai commented on secp256k1-sys/src/lib.rs:
```diff
@@ -365,7 +386,7 @@ unsafe fn strlen(mut str_ptr: *const c_char) -> usize {
 /// In case the type is empty this trait will give a NULL pointer, which should be handled in C.
 /// 
-pub(crate) trait CPtr {
+pub trait CPtr {
```
In a late PR I'll try move this to `secp256k1` and not `-sys` and make it `pub(crate)` again.

stevenroose replied:
Is this not used in -sys? In that case I can move it.. The idea is to try and reduce the times we need to bump a -sys version. So I think better make breaking changes now.

stevenroose replied:
Hmm, seems to be used in macros as well. Yeah I think I'd prefer to merge this as is :)

elichai replied:
Merge != release.
let's not change this PR any more so it can get merged at some point :)

elichai commented:
ACK code review.
ACK reproduced the same changes by the script.
0093a5d283fdd2a88fa33f314346a2c71d079b01

I'm still not so sure about this script but I guess that's the best we can do right now.

apoelstra commented on secp256k1-sys/src/lib.rs:
```diff
@@ -267,6 +287,7 @@ extern "C" {
 // Returns: a newly created context object.
 //  In:      flags: which parts of the context to initialize.
 pub unsafe extern "C" fn secp256k1_context_create(flags: c_uint) -> *mut Context {
+    use std::mem;
```
should this be `core::mem`?

apoelstra replied:
Oh, no, I see, this whole function is `std` gated

apoelstra commented on secp256k1-sys/vendor-libsecp.sh:
```diff
@@ -0,0 +1,68 @@
+
+cd "$PARENT_DIR" || exit 1
+rm -rf "$DIR"
+git clone git@github.com:bitcoin-core/secp256k1.git "$DIR"
```
Can't you just clone over https? This does not work for me because it wants a ssh key.

stevenroose replied:
oh, yeah that makes sense.

stevenroose replied:
I pushed an extra commit.

real-or-random commented:
needs rebase :/
